### ISSUE TITLE
Buggy interplay between ``tf.vectorized_map`` and ``tn.contractors.*``

### ISSUE THREAD

**Issue by refraction-ray** [google/TensorNetwork#940]
When a function ``f`` has, say ``tn.contractors.optimal()`` inside, the results of the function is of course right, but the results for ``tf.vectorized_map(f, batched_input)`` might be wrong. 
See a full demo script at  https://gist.github.com/refraction-ray/3a24e1ca09d9c2850d62b40cbf57946c. 
The results are always correct for individual input (jitted or not) and for ``jax.vmap`` interface but become wrong for ``tf.vectorized_map`` interface. I suspect the problem may be related to the interplay of ``tf.vectorized_map`` and the path solver provided by opteinsum, though I cannot see why they are entangled.

The output for the above demo script in tensorflow case is
```bash
$ python vectorized_map_bug.py tensorflow

3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34)
[GCC 7.3.0]
tensornetwork version: 0.4.5
tensorflow version: 2.5.0
tf.Tensor((-0.4161468+0.90929735j), shape=(), dtype=complex64)
tf.Tensor((-0.65364355-0.7568025j), shape=(), dtype=complex64)
jitted
tf.Tensor((-0.4161468+0.90929735j), shape=(), dtype=complex64)
tf.Tensor((-0.65364355-0.7568025j), shape=(), dtype=complex64)
mapfn
tf.Tensor([-0.4161468 +0.90929735j -0.65364355-0.7568025j ], shape=(2,), dtype=complex64)
vectorized map
tf.Tensor([0.99999994+0.j 0.99999994+0.j], shape=(2,), dtype=complex64)
```

**Comment by mganahl** [comment]
Thanks, I'll take a look!

**Comment by refraction-ray** [comment]
Actually the bug is not necessarily related to ``tn.contractors.*``. I found the same issue persists, even we only use ``contract_between`` in the function:

```python
def u0(inputs):
    """
    compute the following tensor

    node0 -        -       -
      |      gate1   gate2
    node1 -        -       -

    gate1 = gate2 = expm(i*inputs*zz)
    """
    nodes = []
    nodes.append(
        tn.Node(
            np.array(
                [
                    [1.0],
                    [0.0],
                ],
                dtype=npdtype,
            ),
        ),
    )
    bk = nodes[0].backend
    nodes.append(
        tn.Node(
            np.array(
                [
                    [1.0],
                    [0.0],
                ],
                dtype=npdtype,
            ),
        )
    )
    nodes[0][1] ^ nodes[1][1]
    gate1 = tn.Node(
        bk.reshape(
            bk.cos(
                inputs[0],
            )
            * ii
            + bk.sin(inputs[0]) * 1.0j * zz,
            [2, 2, 2, 2],
        )
    )
    gate2 = tn.Node(
        bk.reshape(
            bk.cos(inputs[0]) * ii + bk.sin(inputs[0]) * 1.0j * zz,
            [2, 2, 2, 2],
        )
    )
    nodes.append(gate1)
    nodes.append(gate2)
    gate1[2] ^ nodes[0][0]
    gate1[3] ^ nodes[1][0]
    gate2[2] ^ gate1[0]
    gate2[3] ^ gate1[1]
    s = nodes[0] @ nodes[1]
    s = gate1 @ s
    s = gate2 @ s
    # nn = tn.contractors.optimal(nodes, [gate2[0], gate2[1]])
    return s.tensor[0, 0]


```

**Comment by mganahl** [comment]
I had a look at the script you linked. The following things would be helpful in order to analyse the problem:
1. Could you try to reduce the program to a minimal reproducing program (change to explicitly use tensorflow backend, remove unneeded code, aso)?
2. Explicitly state what the input to the program is (e.g. what types, what dimensions of arrays, aso)
3.  Explicitly state what the program is supposed to compute.
thanks!

**Comment by refraction-ray** [comment]
@mganahl , thanks for your help. 

1. Actually I have try my best to keep the script a minimal to reproduce the bug (smaller network contraction could have no bug), and the more compact script can be found below with only tf backend support:

```python
import tensornetwork as tn
import numpy as np
import tensorflow as tf

tn.set_default_backend("tensorflow")
npdtype = np.complex64
ii = np.eye(4).astype(npdtype)
zz = np.array([[1, 0, 0, 0], [0, -1, 0, 0], [0, 0, -1, 0], [0, 0, 0, 1]], dtype=npdtype)
inputs = np.array([[1.0], [2.0]], dtype=npdtype)


def u0(inputs):
    """
    compute the following tensor

    node0 -        -       -
      |      gate1   gate2
    node1 -        -       -

    gate1 = gate2 = expm(i*inputs*zz)
    """
    nodes = []
    nodes.append(
        tn.Node(
            np.array(
                [
                    [1.0],
                    [0.0],
                ],
                dtype=npdtype,
            ),
        ),
    )
    nodes.append(
        tn.Node(
            np.array(
                [
                    [1.0],
                    [0.0],
                ],
                dtype=npdtype,
            ),
        )
    )
    nodes[0][1] ^ nodes[1][1]
    gate1 = tn.Node(
        tf.reshape(
            tf.math.cos(inputs[0]) * ii + tf.math.sin(inputs[0]) * 1.0j * zz,
            [2, 2, 2, 2],
        )
    )
    gate2 = tn.Node(
        tf.reshape(
            tf.math.cos(inputs[0]) * ii + tf.math.sin(inputs[0]) * 1.0j * zz,
            [2, 2, 2, 2],
        )
    )
    nodes.append(gate1)
    nodes.append(gate2)
    gate1[2] ^ nodes[0][0]
    gate1[3] ^ nodes[1][0]
    gate2[2] ^ gate1[0]
    gate2[3] ^ gate1[1]
    s = nodes[0] @ nodes[1]
    s = gate1 @ s
    s = gate2 @ s
    return s.tensor[0, 0]


print(u0(inputs[0]))
print(u0(inputs[1]))
print(tf.vectorized_map(u0, inputs))
```

2. the program we use is ``u0``, whose input  is a shape [1] complex64 array,
3. and it computes a tensor contraction with the digram shown in u0's docstring while returns the first component of the final tensor. (It corresponds to a quantum circuit simulation with circuit parameters given by ``inputs`` of ``u0`` and returns the first amplitude of the final wavefunction)

**Comment by mganahl** [comment]
thanks! I have tracked down the problem to our tensordot2.py module, a wrapper around tf.tensordot which is a bit more efficient than tensorflow's tensordot function. Since the this module is relatively old, I think we may be able to actually remove it, and simply use tf.tensordot, which fixes the bug. I've prepared a PR with the fix, should be in soon! Thanks again for pointing this out!

**Comment by refraction-ray** [comment]
wow, thanks so much for the quick response! 
``tensordot2`` indeed is the source of the problem, with plain ``tf.tensordot``, the demo script works as expected!

**Comment by mganahl** [comment]
digging deeper, this is in fact a tensorflow bug. Filed a bug report here https://github.com/tensorflow/tensorflow/issues/52148.

**Comment by refraction-ray** [comment]
Even after replacing `tensordot2` with `tensordot`, the buggy interplay between `tf.vectorized_map` and possibly `tf.matmul` remains for gradient evaluation (forward evaluation is ok). See the demo below

```python
import tensornetwork as tn
import numpy as np
import tensorflow as tf

tn.set_default_backend("tensorflow")
npdtype = np.complex64
x = np.array([[0, 1], [1, 0]], dtype=npdtype)
i2 = np.eye(2).astype(npdtype)
s2 = np.sqrt(2.0)
h = np.array([[1 / s2, 1 / s2], [1 / s2, -1 / s2]], dtype=npdtype)
inputs = np.array([[1.0], [2.0]], dtype=npdtype)


def u1(inputs):
    """
    compute the following tensor, as expectation of X0 on a circuit

    0 - H - rx(inputs) - X - rx*(inputs) - H - 0
    0 - H - rx(inputs) - rx*(inputs) - H -0

    the results is trivially 1, and the gradient wrt. inputs should be trivially zero
    """
    nodes = []
    # the first qubit below
    nodes.append(
        tn.Node(
            np.array(
                [1.0, 0.0],
                dtype=npdtype,
            ),
        ),
    )

    nodes.append(tn.Node(h))
    nodes.append(
        tn.Node(tf.math.cos(inputs[0]) * i2 + tf.math.sin(inputs[0]) * 1.0j * x)
    )
    nodes.append(tn.Node(x))
    nodes.append(
        tn.Node(tf.math.cos(inputs[0]) * i2 - tf.math.sin(inputs[0]) * 1.0j * x)
    )
    nodes.append(tn.Node(h))
    nodes.append(
        tn.Node(
            np.array(
                [1.0, 0.0],
                dtype=npdtype,
            ),
        ),
    )
    # the second qubit below
    nodes.append(
        tn.Node(
            np.array(
                [1.0, 0.0],
                dtype=npdtype,
            ),
        ),
    )

    nodes.append(tn.Node(h))
    nodes.append(
        tn.Node(tf.math.cos(inputs[0]) * i2 + tf.math.sin(inputs[0]) * 1.0j * x)
    )
    nodes.append(
        tn.Node(tf.math.cos(inputs[0]) * i2 - tf.math.sin(inputs[0]) * 1.0j * x)
    )
    nodes.append(tn.Node(h))
    nodes.append(
        tn.Node(
            np.array(
                [1.0, 0.0],
                dtype=npdtype,
            ),
        ),
    )
    # connect and contract!
    nodes[0][0] ^ nodes[1][0]
    for i in range(1, 6):
        nodes[i][1] ^ nodes[i + 1][0]
    nodes[7][0] ^ nodes[8][0]
    for i in range(8, 12):
        nodes[i][1] ^ nodes[i + 1][0]
    s = nodes[0] @ nodes[1]
    for i in range(2, 13):
        s = tn.contract_between(s, nodes[i], allow_outer_product=True)
    return tf.math.real(s.tensor)


def vag(inputs):
    with tf.GradientTape() as tape:
        tape.watch(inputs)
        loss = u1(inputs)
    grad = tape.gradient(loss, inputs)
    return loss, grad


inputs = tf.constant(inputs)
print("vag ", vag(inputs[0]))
print("vag ", vag(inputs[1]))
print("vm_vag ", tf.vectorized_map(vag, inputs))
# the value is correct while the gradient is wrong
```

I guess the reason behind it is that the backward evaluation on `tf.contract_between(outer_product=True)` still involves `tf.matmul` op? Can we have a workaround to get correct gradient evaluation in such a case?

 After all, it may take too long for tensorflow to fix this bug, especially considering that complex valued op is not in high priority for ML users :(

**Comment by mganahl** [comment]
Thanks @refraction-ray, I'll take a look.

**Comment by refraction-ray** [comment]
The reason may be that the grad of `matmul` involves the `transpose_a` and `transpose_b` attr, see https://github.com/tensorflow/tensorflow/blob/3fbcdd010b4290c05f0c72a9b23d97664fbb595e/tensorflow/python/ops/math_grad.py#L1721-L1751.

And by directly hacking the tensorflow source from https://github.com/tensorflow/tensorflow/blob/3fbcdd010b4290c05f0c72a9b23d97664fbb595e/tensorflow/python/ops/math_grad.py#L1740-L1741 to
```python
bt = array_ops.transpose(b)
at = array_ops.transpose(a)
grad_a = gen_math_ops.mat_mul(grad, bt)
grad_b = gen_math_ops.mat_mul(at, grad)
```
The bug indeed vanishes. Just don't know the elegant way to monkey patch the grad function of native tf op.
